presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,181 @@
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusClient
+{
+    private static PrometheusConnectorConfig config;
+    static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
+    private static final OkHttpClient httpClient = new OkHttpClient.Builder().build();
+    private final Supplier<Map<String, Object>> tableSupplier;
+    private TypeManager typeManager;
+
+    @Inject
+    public PrometheusClient(PrometheusConnectorConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
+            throws URISyntaxException
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(metricCodec, "metricCodec is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        tableSupplier = Suppliers.memoizeWithExpiration(metricsSupplier(metricCodec, getPrometheusMetricsURI(config)),
+                (long) config.getCacheDuration().getValue(), config.getCacheDuration().getUnit());
+        this.typeManager = typeManager;
+        this.config = config;
+    }
+
+    public Set<String> getSchemaNames()
+    {
+        return ImmutableSet.of("default");
+    }
+
+    private URI getPrometheusMetricsURI(PrometheusConnectorConfig config)
+            throws URISyntaxException
+    {
+        // endpoint to retrieve metric names from Prometheus
+        URI uri = config.getPrometheusURI();
+        return new URIBuilder()
+                .setScheme(uri.getScheme())
+                .setHost(uri.getAuthority())
+                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
+                .build();
+    }
+
+    public Set<String> getTableNames(String schema)
+    {
+        requireNonNull(schema, "schema is null");
+        String status = "";
+        if (schema.equals("default")) {
+            status = (String) tableSupplier.get().get("status");
+            //TODO prometheus warnings (success|error|warning) could be handled separately
+            if (status.equals("success")) {
+                List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+                if (tableNames == null) {
+                    return ImmutableSet.of();
+                }
+                return ImmutableSet.copyOf(tableNames);
+            }
+        }
+        throw new PrestoException(PROMETHEUS_TABLES_METRICS_RETRIEVE_ERROR, String.format("Prometheus did no return metrics list (table names): %s", status));
+    }
+
+    public PrometheusTable getTable(String schema, String tableName)
+    {
+        requireNonNull(schema, "schema is null");
+        requireNonNull(tableName, "tableName is null");
+        if (schema.equals("default")) {
+            List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+            if (tableNames == null) {
+                return null;
+            }
+            if (!tableNames.contains(tableName)) {
+                return null;
```

Comment:
return `Optional`

Reply:
I continue to resist this change as unless TableNotFoundException supports Supplier it's not a valuable change to functional style.